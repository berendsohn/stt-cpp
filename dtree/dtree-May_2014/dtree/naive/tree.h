// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
