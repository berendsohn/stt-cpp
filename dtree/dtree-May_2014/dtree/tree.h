// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
