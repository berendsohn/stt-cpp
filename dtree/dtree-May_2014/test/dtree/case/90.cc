#define DTREE_CONFIG EndTree<WithEvert<WithEvert<WithAncAggr<dtree::DpMax<double>, dtree::DwGreater_<double>, WithDescValue<dtree::DpAdd<double>, WithAncAggr<dtree::Count<int>, dtree::Index_<int>, Begin > > > > > >
#include "test/dtree/tree.cc"
