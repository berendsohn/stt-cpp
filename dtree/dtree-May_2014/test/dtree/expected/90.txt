<<<
  1: vL-211R-323F0 -> 9
  2: vL-245R-377F0
  3: vL-317R-405F0 -> 5
  4: vL168R-134F0 -> 3
  5: vL85R-8F0 -> 12
  6: vL-141R-254F0 -> 7
  7: vL-75R-204F0 -> 1
  8: vL-15R121F0 -> 15
  9: vL-66R40F0 -> 2
  10: vL-155R43F0 -> 7
  11: vL117R-277F0 -> 12
  12: vL204R-449F0 -> 15
  13: vL56R-314F0 -> 10
  14: vL37R-205F0 -> 7
  15: vL-204R-89F0 -> 9
>>>
<<<
    /9: a2 vL179R417F0 aL0R0 f0 f1
      \1: a1 vL-145R-363F0 aL0R0 f0 f0
  @2: a1 vL-245R-377F0 aL0R0 f0 f1

 9
    /3: a4 vL-485R-271F0 aL521R397 f0 f1
        /15: a1 vL-408R360F0 aL0R0 f0 f0
      \12: a3 vL521R-44F0 aL0R441 f1 f0
        \5: a1 vL-119R441F0 aL0R0 f0 f1
  \4: a5 vL234R-174F0 aL36R126 f0 f0

 1
    /6: a1 vL-66R-50F0 aL0R0 f1 f0
  \7: a2 vL136R119F0 aL0R0 f0 f1

 15
  \8: a1 vL189R210F0 aL0R0 f0 f1

 7
    /13: a1 vL211R-357F0 aL0R0 f1 f0
  \10: a2 vL-80R247F0 aL211R0 f0 f1

 12
  \11: a1 vL-87R172F0 aL0R0 f1 f1

 7
  \14: a1 vL112R-1F0 aL0R0 f0 f0
>>>
